apis/ninedof/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+#![no_std]
+
+use core::cell::Cell;
+use libtock_platform::{
+    share, share::Handle, subscribe::OneId, DefaultConfig, ErrorCode, Subscribe, Syscalls, Upcall,
+};
+
+pub struct NineDof<S: Syscalls>(S);
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NineDofData {
+    pub x: i32,
+    pub y: i32,
+    pub z: i32,
+}
+
+fn sqrt(n: f64) -> f64 {
+    /// Newton-Raphson method for sqrt
+    const EPSILON: f64 = 0.00001;
+    let mut x: f64 = n;
+    let mut y: f64 = 1.0;
+
+    while x - y > EPSILON {
+        x = (x + y) / 2.0;
+        y = n / x;
+    }
+
+    x
+}
+
+impl<S: Syscalls> NineDof<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    /// Initiate a accelerometer measurement.
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_accelerometer() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_ACCELEROMETER, 0, 0).to_result()
+    }
+
+    /// Initiate a magnetometer measurement.
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_magnetometer() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_MAGNETOMETER, 0, 0).to_result()
+    }
+
+    /// Initiate a gyroscope measurement.
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_gyro() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_GYRO, 0, 0).to_result()
+    }
+
+    /// Unregister an events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Register an events listener
+    pub fn register_listener<'share, F: Fn(NineDofData)>(
+        listener: &'share NineDofListener<F>,
+        subscribe: Handle<Subscribe<'share, S, DRIVER_NUM, 0>>,
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Initiate a synchronous accelerometer measurement.
+    /// Returns Ok(accelerometer_value) if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_accelerometer_sync() -> Result<NineDofData, ErrorCode> {
+        let data_cell: Cell<Option<NineDofData>> = Cell::new(None);
+        let listener = NineDofListener(|data| {
+            data_cell.set(Some(data));
+        });
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::read_accelerometer() {
+                    while data_cell.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
+
+        match data_cell.get() {
+            None => Err(ErrorCode::Busy),
```

Comment:
```suggestion
            None => Err(ErrorCode::Busy),
```

This is one of the return values of the first two system calls. Same for the next ones.

Reply:
I don't understand what should I return here, ```Err(ErrorCode::Busy)``` or what error should I use?

Reply:
Return the error that `subscribe` or `read_accelerometer` return.